File: 2-DataStructures/6-Hash-Tables/hash-tables.js
// Hash Tables - Definition: A data structure that stores values by key, unlike an array that stores values by index.
// What more: Hash tables are fast for all of the following operations: finding values, adding new values, and removing values!
//  - Hash tables are collections of key-value pairs
//  - Hash tables can find values quickly given a key
//  - Hash tables can add new key-values quickly

let user = {
  age: 54,
  name: "Kylie",
  magic: true,
  scream: function () {
    console.log("ahhhhh!");
  },
};

user.age; // O(1)
user.spell = "abra kadabra"; // O(1)
user.scream(); // O(1)

// implementing hash tables:
class HashTable {
  constructor(size) {
    this.data = new Array(size);
  }

  _hash(key) {
    let hash = 0;
    for (let i = 0; i < key.length; i++) {
      hash = (hash + key.charCodeAt(i) * i) % this.data.length;
    }
    return hash;
  }

  set(key, value) {
    let address = this._hash(key);
    if (!this.data[address]) {
      this.data[address] = [];
    }
    this.data[address].push([key, value]);
    return this.data;
  }

  get(key) {
    let address = this._hash(key);
    const currentBucket = this.data[address];
    console.log(currentBucket);

    if (currentBucket.length) {
      for (let i = 0; i < currentBucket.length; i++) {
        if (currentBucket[i][0] === key) {
          return currentBucket[i][1];
        }
      }
    }
    return undefined;
  }

  // keys() {
  //   const keysArray = [];
  //   for (let i=0; i < this.data.length; i++) { // fix it
  //     if (this.data[i]) {
  //       keysArray.push(this.data[i][0][0]);
  //     }
  //   }
  //   return keysArray;
  // }

  // keys with collision prevention
  keys() {
    if (!this.data.length) {
      return undefined
    }
    let result = []
    // loop through all the elements
    for (let i = 0; i < this.data.length; i++) {
        // if it's not an empty memory cell
        if (this.data[i] && this.data[i].length) {
          // but also loop through all the potential collisions
          if (this.data.length > 1) {
            for (let j = 0; j < this.data[i].length; j++) {
              result.push(this.data[i][j][0])
            }
          } else {
            result.push(this.data[i][0])
          } 
        }
    }
    return result; 
  }
}

const myHashTable = new HashTable(10);
myHashTable.set("grapes", 1000); // [['grapes', 1000]]
myHashTable.set("bananas", 20);
myHashTable.set("apples", 2);
console.log(myHashTable);
console.log(myHashTable.get("grapes"));
console.log(myHashTable.keys());

File: 2-DataStructures/5-Hash-Tables/hash-tables.py
# Hash Tables: are used to store key-value pairs.
# Hash Tables are fast for finding, adding and removing values.
# Hash Tables are unordered.
# Hash Tables are called Dictionaries in Python.
# Operations: Insert, Lookup, Delete, Search

class HashTable:
    def __init__(self, size):
        self.data = [None] * size

    def _hash(self, key):
        hash_value = 0
        for i in range(len(key)):
            hash_value = (hash_value + ord(key[i]) * i) % len(self.data)
        return hash_value

    def set(self, key, value):
        address = self._hash(key)
        if not self.data[address]:
            self.data[address] = []
        self.data[address].append([key, value])
        return self.data

    def get(self, key):
        address = self._hash(key)
        current_bucket = self.data[address]
        print(current_bucket)

        if current_bucket:
            for i in range(len(current_bucket)):
                if current_bucket[i][0] == key:
                    return current_bucket[i][1]
        return None

    def keys(self):
        if not self.data:
            return None
        result = []
        for i in range(len(self.data)):
            if self.data[i] and len(self.data[i]) > 0:
                if len(self.data[i]) > 1:
                    for j in range(len(self.data[i])):
                        result.append(self.data[i][j][0])
                else:
                    result.append(self.data[i][0][0])
        return result


myHashTable = HashTable(10)
myHashTable.set("grapes", 1000)  # [['grapes', 1000]]
myHashTable.set("bananas", 20)
myHashTable.set("apples", 2)
print(myHashTable.data)
print(myHashTable.get("grapes"))
print(myHashTable.keys())
